protocols/smtp/src/main/java/org/apache/james/protocols/smtp/core/AbstractSenderAuthIdentifyVerificationRcptHook.java

```diff
@@ -46,29 +47,23 @@
     public HookResult doRcpt(SMTPSession session, MaybeSender sender, MailAddress rcpt) {
         if (session.getUser() != null) {
             String authUser = (session.getUser()).toLowerCase(Locale.US);
-            MailAddress senderAddress = (MailAddress) session.getAttachment(
+            MaybeSender senderAddress = (MaybeSender) session.getAttachment(
                     SMTPSession.SENDER, ProtocolSession.State.Transaction);
-            String username = retrieveSender(sender, senderAddress);
+            Optional<String> username = sender.asOptional().map(this::getUser);
             
             // Check if the sender address is the same as the user which was used to authenticate.
             // Its important to ignore case here to fix JAMES-837. This is save todo because if the handler is called
             // the user was already authenticated
-            if ((senderAddress == null)
-                || (!authUser.equalsIgnoreCase(username))
-                || (!isLocalDomain(senderAddress.getDomain()))) {
+            if (senderAddress == null
+                || senderAddress.isNullSender()
```

Comment:
the mix of tests on senderAdress and username is weird.
Could you try to sort things in a way that make sense ?
Why even compute username before checking senderAddress ?

Reply:
I extracted the various part of this 'if' in separated methods.